File: apps/ios/Barns/BarnsApp/Presentation/Catalog/CatalogListViewModel.swift
import Combine

/// Loads the supporting, official read-only catalog list through a use case.
@MainActor
final class CatalogListViewModel: ObservableObject {
    enum State: Equatable {
        case loading
        case loaded([CatalogPresentationItem])
        case failed(String)
    }

    @Published private(set) var state: State = .loading

    private let getCatalogItemsUseCase: GetCatalogItemsUseCase

    init(getCatalogItemsUseCase: GetCatalogItemsUseCase) {
        self.getCatalogItemsUseCase = getCatalogItemsUseCase
    }

    func load() async {
        state = .loading
        do {
            let items = try await getCatalogItemsUseCase.execute()
            state = .loaded(items.map(CatalogPresentationItem.init))
        } catch {
            state = .failed("Unable to load the catalog. Please try again.")
        }
    }
}

File: apps/android/app/src/main/java/com/barns/app/presentation/catalog/CatalogListViewModel.kt
package com.barns.app.presentation.catalog

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.barns.app.domain.usecase.catalog.GetCatalogItemsUseCase
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch

/** Loads the supporting, official read-only catalog list through a use case. */
class CatalogListViewModel(
    private val getCatalogItemsUseCase: GetCatalogItemsUseCase,
) : ViewModel() {
    sealed interface State {
        object Loading : State
        data class Loaded(val items: List<CatalogPresentationItem>) : State
        data class Failed(val message: String) : State
    }

    private val _state = MutableStateFlow<State>(State.Loading)
    val state: StateFlow<State> = _state.asStateFlow()

    fun load() {
        _state.value = State.Loading
        viewModelScope.launch {
            runCatching { getCatalogItemsUseCase.execute() }
                .onSuccess { items ->
                    _state.value = State.Loaded(items.map(CatalogPresentationItem::from))
                }
                .onFailure {
                    _state.value = State.Failed("Unable to load the catalog. Please try again.")
                }
        }
    }
}
